esp32_ulp/preprocess.py

```diff
@@ -2,6 +2,46 @@
 from .util import split_tokens
 
 
+class RTC_Macros:
+    @staticmethod
+    def READ_RTC_REG(args):
+        rtc_reg, low_bit, bit_width = args
+        args = (
+            rtc_reg,
+            '%s + %s - 1' % (low_bit, bit_width),
+            low_bit
+        )
+        return '\treg_rd ' + ', '.join(args)
+
+    @staticmethod
+    def WRITE_RTC_REG(args):
+        rtc_reg, low_bit, bit_width, value = args
+        args = (
+            rtc_reg,
+            '%s + %s - 1' % (low_bit, bit_width),
+            low_bit,
+            value
+        )
+        return '\treg_wr ' + ', '.join(args)
+
+    @staticmethod
+    def READ_RTC_FIELD(args):
+        rtc_reg, low_bit = args
+        return RTC_Macros.READ_RTC_REG((rtc_reg, low_bit, 1))
+
+    @staticmethod
+    def WRITE_RTC_FIELD(args):
+        rtc_reg, low_bit, value = args
+        return RTC_Macros.WRITE_RTC_REG((rtc_reg, low_bit, 1, value + ' & 1'))
+
+    @staticmethod
+    def eval_expr(expr):
+        try:
+            return eval(expr)
+        except:
+            return expr
```

Comment:
unused? if so, please remove.

Reply:
Sneaky method: hiding in plain sight! I will remove.
(Originally I did have expression evaluation in the preprocessor - but felt it was better in the assembler, made the preprocessor simpler and by moving it there it allowed for that feature to exist already before the preprocessor).